### PR TITLE
Resources outside the `*kubernetes.io` namespace are integers and cannot be over-committed.

### LINKED ISSUE #50473: Reduce code changes needed to support a new resource name.
/kind feature

Any resource without a domain name or with `kubernetes.io/` domain name should continue to be treated as first class. Everything else could be treated as integer resources, until a future API allows specifying whether a resource is integral/fractional.

[Opaque Integer Resources](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#opaque-integer-resources-alpha-feature) (also called OIR, alpha in v1.5.0), is implemented by special-casing resources with a certain prefix. Since the default scheduler currently handles all resources except CPU as integers, this could be codified as default behavior for resources that do not belong to the `kubernetes.io/` namespace. This would reduce the proliferation of very similar code in admission control, scheduler predicates, kubelet, kubectl, etc. when adding support for additional classes of resources, such as resources advertised by extensions. The initial use case is the upcoming alpha device plugin interface. Resources advertised via device plugins will share the `extensions.kubernetes.io/` namespace.

**Related issues:**
* [feature: extension to support new compute resources](https://github.com/kubernetes/features/issues/368)
* [proposal: hardware accelerators](https://github.com/kubernetes/community/pull/844)

cc @vishh @renaudwastaken @thockin @davidopp

### PR DESCRIPTION
**Pull request by ConnorDoyle**
**What this PR does / why we need it**:

Fixes #50473 

Rationale: since the scheduler handles all resources except CPU as integers, that could just be the default behavior for namespaced resources.

cc @RenaudWasTaken @vishh 

**Release note**:

```release-note
Resources outside the `*kubernetes.io` namespace are integers and cannot be over-committed.
```

### CODE REVIEW

**Comment by ConnorDoyle** [comment]
+ @davidopp @bsalamat @timothysc (sig-scheduling, reviewed original OIR patches)
+ @thockin (commented on ugliness of OIR prefix in reviews)

**Comment by vishh** [comment]
I'd split compute resources into three categories:

1. First class resources - These resources are expected to be available across all kube deployments with predictable behavior. They typically include special handling across the system. Examples include `cpu`, `memory` and `storage`.
2. Extended Resources - These resources aren't expected to be available across all kube deployments. These resources require special extensions at the node level, and possibly at the scheduler level too. Examples include acclerators, special NICs, etc.
3. Opaque Resources - These resources aren't expected to be available across all kube deployments. These resources do not require any extensions across the system. Examples include Software licences.

I intend to introduce a more feature rich Compute Resource representation API in the future, which can allow for representing both integer and fractional resources across all the three categories mentioned above.

tl;dr; I don't think we should deprecate OIRs. We should rather have well defined user journeys for each of these categories and let users choose which one they want to use.
`2` and `3` may be the same at the scheduler level, but I'm expecting the User Journeys to expose the differences.
cc @steveperry-53

**Comment by ConnorDoyle** [comment]
@vishh that makes a lot of sense to me. So currently the first-class resources live in the default namespace, also available to without fully-qualifying the name. In your mind would it be another namespace that distinguishes extended resources from opaque resources? IIUC that was the function of the proposed `extensions.kubernetes.io/` namespace we talked about.

**Comment by timothysc** [comment]
@ConnorDoyle What's the user story behind this?  Resource accounting?  I keep going back and forth, but it would help me a lot if you had a concrete user story.

**Comment by ConnorDoyle** [comment]
@timothysc here are some (others please chime in if I missed the mark):

* As a Kubelet contributor designing the device plugin subsystem, I want a way to reliably distinguish "opaque" resources from "extended" resources, where extended resources require the Kubelet to take isolation/monitoring/cleanup actions, and opaque resources do not.
* As a device plugin author, I want to advertise new resources for each device managed by my plugin so that users can consume those devices in pod specs.
* As a Kubernetes pod author, I want to ensure my pod is scheduled on a node that has sufficient devices that my workload needs, and for those devices to be appropriately exposed inside my pod's containers.

The device plugin proposal (linked in description) takes a different approach by extending the node spec to have `DeviceCapacity` instead of reusing the existing capacity and allocatable maps. So, this is a mini-counter proposal about how device resources are accounted.

File: pkg/api/helper/helpers.go [hunk 1]
```diff
@@ -113,9 +113,18 @@ func IsStandardContainerResourceName(str string) bool {
 }
 
 // IsOpaqueIntResourceName returns true if the resource name has the opaque
-// integer resource prefix.
+// integer resource prefix, or it is another unknown resource name not in the
+// default namespace.
 func IsOpaqueIntResourceName(name api.ResourceName) bool {
-	return strings.HasPrefix(string(name), api.ResourceOpaqueIntPrefix)
+	return IsCustomNamespacedResourceName(name)
```
Comment (ConnorDoyle) [inline pkg/api/helper/helpers.go]: We could alternatively do something like:

```go
func IsIntResourceName(name api.ResourceName) bool {
    return IsOpaqueIntResourceName(name) || IsExtendedResourceName(name)
}
```

And then update current uses of `IsOpaqueIntResourceName()` in the scheduler to use `IsIntResourceName()` instead. This would preserve the existing OIR prefix behavior and resources with the new `"extensions.kubernetes.io/"` prefix ride along with the existing scheduler logic for OIRs.

Comment (vishh) [inline pkg/api/helper/helpers.go]: @davidopp I like this change in general. It allows for exposing arbitrary compute resources in all domains.

**Comment by vishh** [comment]
Other resource APIs like Quota, Limit Range, etc., aren't extensible yet. So this PR will lead to feature requests for Quota and Limit range API extensions @davidopp

File: pkg/api/helper/helpers.go (hunk 1 above)
Comment (ConnorDoyle) [inline pkg/api/helper/helpers.go]: @vishh do you mean you like the patch as written, or the change suggested in this comment?

File: pkg/api/helper/helpers.go [hunk 2]
```diff
@@ -113,9 +113,18 @@ func IsStandardContainerResourceName(str string) bool {
...
+}
+
+// Returns true if the resource name is not in the kubernetes.io/ namespace.
+// Note: partially-qualified (unprefixed) names are assumed by default to be
+//       in the kubernetes.io/ namespace.
+func IsCustomNamespacedResourceName(name api.ResourceName) bool {
+	return strings.Contains(string(name), "/") &&
```
Comment (davidopp) [inline pkg/api/helper/helpers.go]: since api.ResourceDefaultNamespacePrefix ends with a slash, why do you need this line?

Comment (davidopp) [inline pkg/api/helper/helpers.go]: oh, never mind

File: pkg/api/helper/helpers.go (hunk 1 above)
Comment (vishh) [inline pkg/api/helper/helpers.go]: I don't see the point in adding more and more special cased prefixes in the API. 
I like the current proposal. Any resources without a domain name or with `kubernetes.io` domain name will be treated as first class resources.
Everything else will be treated as integer resources until we have a better API representation for resources that allows to represent non-integer resources.

**Comment by thockin** [comment]
This needs an associated issue and a `closes #12345` in the FIRST COMMENT, or it can not be approved.

**Comment by ConnorDoyle** [comment]
Updated to fix unit test failure.

**Comment by vishh** [comment]
@ConnorDoyle This is necessary for proposal - https://github.com/kubernetes/community/pull/844

Can you also file an issue that talks about the need for this change as per @thockin request in https://github.com/kubernetes/kubernetes/pull/48922#issuecomment-321459689 ?

**Comment by ConnorDoyle** [comment]
@thockin @vishh filed issue #50473 for this

File: pkg/api/helper/helpers.go (hunk 1 above)
Comment (jiayingz) [inline pkg/api/helper/helpers.go]: Would this also include "alpha.kubernetes.io/xxx" or ResourceStorageScratch or ResourceStorageOverlay? Generally I feel it is probably safer to start with:
func IsIntResourceName(name api.ResourceName) bool {
    return IsOpaqueIntResourceName(name) || IsExtendedResourceName(name)
}
to make sure we don't break the current resource behavior.

Comment (ConnorDoyle) [inline pkg/api/helper/helpers.go]: @jiayingz good catch; the predicate as written would match `alpha.kubernetes.io/` resources.

Another option could be to match the substring instead of the prefix.

```go
func IsCustomNamespacedResourceName(name api.ResourceName) bool {
	return strings.Contains(string(name), "/") &&
		!strings.Contains(string(name), api.ResourceDefaultNamespacePrefix)
}
```

Comment (jiayingz) [inline pkg/api/helper/helpers.go]: Hmm, would that work if device plugin resource also contains ResourceDefaultNamespacePrefix?

Comment (ConnorDoyle) [inline pkg/api/helper/helpers.go]: Do we plan to have additional `alpha.kubernetes.io/xxx` resources that are fractional? This patch would already treat them as integer resources.

Comment (jiayingz) [inline pkg/api/helper/helpers.go]: Probably not but they will get surprise when they do. What do you think we explicitly check !strings.HasPrefix(string(name), api.ResourceDefaultNamespacePrefix) && !strings.HasPrefix(string(name), "alpha." + api.ResourceDefaultNamespacePrefix) && !strings.HasPrefix(string(name), "beta." + api.ResourceDefaultNamespacePrefix)?

Comment (ConnorDoyle) [inline pkg/api/helper/helpers.go]: Fine with that, but it does force a developer adding support for a new resource like `alpha.kubernetes.io/foo` to touch a lot of code. At this point maybe we could defer to @davidopp or @bsalamat for input from the sig-scheduling angle.

Comment (vishh) [inline pkg/api/helper/helpers.go]: Reserving `kubernetes.io` domain for first class resources is what I assumed this PR is introducing. If that's true, `alpha.kubernetes.io` is a subdomain of `kubernetes.io` and it should also be considered to be a first class resource. 
Any compute resource under the `kubernetes.io` or the default domain has to be explicitly added to the system across the scheduler, admission, node, quota, limitrange, UI, etc.

For now, all non-first class resources are considered to be integer resources without any QoS semantics (requests == limits).

Comment (vishh) [inline pkg/api/helper/helpers.go]: The issue of extensibility across quota, limitrange, UI, etc., is beyond just the APIs. I'd recommend deferring it since it requires changes to the rest of the system as well.

File: pkg/api/helper/helpers.go [hunk 3]
```diff
@@ -116,9 +116,18 @@ func IsStandardContainerResourceName(str string) bool {
...
+		!strings.HasPrefix(string(name), api.ResourceDefaultNamespacePrefix)
```
Comment (vishh) [inline pkg/api/helper/helpers.go]: Can we check that the domain name is `ResourceDefaultNamespacePrefix` instead of just the prefix.?

File: pkg/api/helper/helpers.go (hunk 1 above)
Comment (ConnorDoyle) [inline pkg/api/helper/helpers.go]: @jiayingz pointed out that extensions.kubernetes.io is also part of the domain.

**Comment by ConnorDoyle** [comment]
Filling in from chat w/ @vishh:

1. Announce deprecation cycle for OIRs beginning with v1.8. Users must opt-in to legacy OIR behavior from v1.8.
1. Update this PR to treat everything not in the `*kubernetes.io/` domain as an integer resource
    1. ...except for OIR-prefixed resources, if enabled via config.
1. Remove legacy OIR behavior entirely in a later release (e.g. v1.9.)

File: pkg/api/helper/helpers.go (hunk 1 above)
Comment (davidopp) [inline pkg/api/helper/helpers.go]: Let me see if I understand. With the suggested change, `alpha.kubernetes.io/*`, `beta.kubernetes.io/*`, `kubernetes.io/*`, and any resource not of the form `xxx/*` will be considered a first class resource. All resources of the form `xxx/*` that does not match `alpha.kubernetes.io/*`, `beta.kubernetes.io/*`, or `kubernetes.io/*` will be considered opaque integer resources (or extended resources, which I admit I haven't read up on, so I don't really know what the desired semantics are for that). Is that correct?

If so, it seems fine to me, assuming it doesn't break any existing resources.

Comment (vishh) [inline pkg/api/helper/helpers.go]: @davidopp Any resource that's of the form `[xxx.]*.kubernetes.io/` or without a domain name is considered to be first class. So `alpha.kubernetes.io` or `foo.bar.kubernetes.io` or `storage.kubernetes.io` are all first class resources.

Everything else with a domain name that's not `kubernetes.io` will be considered as extended resource which are opaque to the scheduler and possibly opaque to the node.
To begin with all these extended resources are considered to be integers.

Comment (davidopp) [inline pkg/api/helper/helpers.go]: Yeah that seems fine. It's really hard to predict how/when a decision like this will bite us later on, but I can't see any reason right now why it might be a problem...

**Comment by ConnorDoyle** [comment]
@davidopp @vishh @thockin updated.

The previous OIR behavior is not opt-in yet. I would like to work on that in a follow-up in order to unblock device plugins.

File: pkg/api/v1/helper/helpers.go [hunk 4]
```diff
@@ -27,6 +27,20 @@ import (
...
+// IsExtendedResourceName returns true if the resource name is not in the
+// default namespace, or it has the opaque integer resource prefix.
+func IsExtendedResourceName(name v1.ResourceName) bool {
+	return IsCustomNamespacedResourceName(name) || IsOpaqueIntResourceName(name)
+}
+
+// Returns true if the resource name is not in the *kubernetes.io/ namespace.
+// Note: partially-qualified (unprefixed) names are assumed by default to be
+//       in the kubernetes.io/ namespace.
+func IsCustomNamespacedResourceName(name v1.ResourceName) bool {
```
Comment (davidopp) [inline pkg/api/v1/helper/helpers.go]: This is just the inverse of the IsDefaultNamespaceResource() function in pkg/api/helper/helpers.go. Can you at least add a comment to both of them so that someone in the future who changes them will ensure they stay in sync?

Comment (ConnorDoyle) [inline pkg/api/v1/helper/helpers.go]: Sorry, I don't think we need both. Will remove one of them. Good catch.

Comment (ConnorDoyle) [inline pkg/api/v1/helper/helpers.go]: Updated

**Comment by thockin** [comment]
Is this right?  Given the design for device plugins, do we believe that all such plugin extensions represent integer-countable resources?  I'll approve, because Vish is on it, but I'd like that spelled out somewhere.

/approve

**Comment by ConnorDoyle** [comment]
@thockin it should be included in the docs for the device plugin feature (at least). Some of the discussion happened on Vish's hardware accelerator [proposal](https://github.com/kubernetes/community/pull/844#discussion_r132525745). To summarize, we didn't want to encode per-resource attributes like fractional-ness and overcommit-ibility in the resource name, and there's no API structure yet for them. Our thought is that this approach has a decent chance of being forward-compatible if one is added.

**Comment by vishh** [comment]
@thockin 

> do we believe that all such plugin extensions represent integer-countable resources?

The plan is to introduce a better API for representing resources (as discussed offline) which would then allow for opening up sharing of extended resources. Until then, extended resources are expected to not allow sharing. If we go with the option of allowing sharing now, then taking that away when a new API pipeline gets introduced will become complicated.
